File: src/leetcode/editor/cn/P5_LongestPalindromicSubstring.java
//给你一个字符串 s，找到 s 中最长的回文子串。 
//
// 
//
// 示例 1： 
//
// 
//输入：s = "babad"
//输出："bab"
//解释："aba" 同样是符合题意的答案。
// 
//
// 示例 2： 
//
// 
//输入：s = "cbbd"
//输出："bb"
// 
//
// 
//
// 提示： 
//
// 
// 1 <= s.length <= 1000 
// s 仅由数字和英文字母组成 
// 
//
// Related Topics 字符串 动态规划 👍 5538 👎 0

package leetcode.editor.cn;

//Java:最长回文子串
//Time:2022-08-05 14:13:35
class P5_LongestPalindromicSubstring {
    public static void main(String[] args) {
        Solution solution = new P5_LongestPalindromicSubstring().new Solution();
        System.out.println(solution.longestPalindrome("cbdd"));

    }

    //leetcode submit region begin(Prohibit modification and deletion)
    class Solution {
        public String longestPalindrome(String s) {
            int last = 0;
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i <= s.length() - 1; i++) {
                String t = s.substring(0, i + 1);
                StringBuilder substr = new StringBuilder(t);
                if (substr.reverse().toString().equals(t)) {
                    last = substr.length();
                    sb = substr;
                }
            }
            return sb.toString();
        }

    }
//leetcode submit region end(Prohibit modification and deletion)

}

File: python/leetcode/editor/cn/P5_LongestPalindromicSubstring.py
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:

    # 求回文半径
    def expand(self, s: str, left: int, right: int) -> int:
        while left >= 0 and right < len(s) and s[left] == s[right]:
            left -= 1
            right += 1
        # 当左右不等或者越界时停止，所以需要减掉2
        return (right - left - 2) // 2

    def longestPalindrome(self, s: str) -> str:
        start, end = 0, -1
        s = '#' + '#'.join(s) + '#'
        # 数组P，存放回文半径
        P = []
        # 记录当前的最右边界
        y = -1
        # 中心点
        c = -1
        # 从i开始向右遍历
        for i in range(len(s)):
            # 如果当前点在最右边界之内
            if y >= i:
                # 拿到i的对称点:i_sym=c-(i-c)=2*c-i
                i_sym = 2 * c - i
                # 是否超过了最远边界,比较i的最远边界（通过对称点获取）和y-i
                min_arm_len = min(P[i_sym], y - i)
                # 从当前位置开始扩展
                cur_arm_len = self.expand(s, i - min_arm_len, i + min_arm_len)
            else:
                # 否则正常扩展
                cur_arm_len = self.expand(s, i, i)
            P.append(cur_arm_len)
            # 更新最远边界
            if i + cur_arm_len > y:
                c = i
                y = c + cur_arm_len
            # 更新结果
            if 2 * cur_arm_len + 1 > end - start:
                start = i - cur_arm_len
                end = i + cur_arm_len
        # 最后start和end一定会落在#上
        return s[start + 1:end + 1:2]


# leetcode submit region end(Prohibit modification and deletion)


print(Solution().longestPalindrome("cbbd"))
